workflow-ui/backstack-android/src/main/java/com/squareup/workflow1/ui/backstack/ViewStateCache.kt

```diff
@@ -75,22 +97,61 @@ internal constructor(
         }
       }
 
+    // Create and install a new SavedStateRegistryOwner for the new view and wire it up to the
+    // view's lifecycle. We need to do this even if the StateRegistryHolder hasn't been restored
+    // yet, the view can ask for it as soon as it's attached.
+    val registryOwner = KeyedStateRegistryOwner.installAsSavedStateRegistryOwnerOn(newView, newKey)
+    currentOwner = registryOwner
+
+    // If the holder hasn't been restored yet, this will be a no-op, but the currentOwner will be
+    // restored by the onRestored callback to the holder constructor.
+    registryHolder.restoreRegistryControllerIfReady(newKey, registryOwner.controller)
     viewStates.remove(newKey)
       ?.let { newView.restoreHierarchyState(it.viewState) }
 
+    // Save both the view state and state registry of the view that's going away, as long as it's
+    // still in the backstack.
     if (oldViewMaybe != null) {
       oldViewMaybe.namedKey.takeIf { hiddenKeys.contains(it) }
         ?.let { savedKey ->
+          // View state
           val saved = SparseArray<Parcelable>().apply {
             oldViewMaybe.saveHierarchyState(this)
           }
           viewStates += savedKey to ViewStateFrame(savedKey, saved)
+
+          // Registry state (note reverse order as saved above)
+          val owner = ViewTreeSavedStateRegistryOwner.get(oldViewMaybe) as KeyedStateRegistryOwner
+          registryHolder.saveRegistryController(savedKey, owner.controller)
         }
     }
 
     pruneKeys(hiddenKeys)
   }
 
+  /**
+   * Must be called whenever the owning view is [attached to a window][View.onAttachedToWindow].
+   * Must eventually be matched with a call to [detachFromParentRegistry].
+   *
+   * Calls [StateRegistryHolder.attachToParentRegistry] – see that method for more information.
+   */
+  public fun attachToParentRegistry(
```

Comment:
Do we need to update the class KDoc to reference these new public methods? (Not a leading question – just mentioning in case it makes sense.)

Reply:
Couldn't hurt.